File: 01/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer writer.Flush()

	line, _ := reader.ReadString('\n')
	parts := strings.Fields(strings.TrimSpace(line))
	R, _ := strconv.Atoi(parts[0])
	B, _ := strconv.Atoi(parts[1])

	W, H := solve(R, B)
	writer.WriteString(fmt.Sprintf("%d %d\n", W, H))
}

// solve находит размеры панели W и H (W >= H) по количеству красных R и синих B плиток
// R = 2*W + 2*H - 4, B = (W-2) * (H-2)
func solve(R, B int) (int, int) {
	sum := (R + 4) / 2

	for d := 1; d*d <= B; d++ {
		if B%d != 0 {
			continue
		}

		// Проверяем оба варианта: (W-2, H-2) = (d, B/d) и (B/d, d)
		for _, w := range []int{B/d + 2, d + 2} {
			h := sum - w
			if (w-2)*(h-2) == B && w >= h {
				return w, h
			}
		}
	}

	return 0, 0
}

File: 01/main.rs
use std::io::{self, BufRead, BufWriter, Write};

fn main() {
    let stdin = io::stdin();
    let mut reader = stdin.lock();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    let mut line = String::new();
    reader.read_line(&mut line).unwrap();
    let parts: Vec<&str> = line.trim().split_whitespace().collect();
    let r: i32 = parts[0].parse().unwrap();
    let b: i32 = parts[1].parse().unwrap();

    let (w, h) = solve(r, b);
    writeln!(writer, "{} {}", w, h).unwrap();
    writer.flush().unwrap();
}

// solve находит размеры панели W и H (W >= H) по количеству красных R и синих B плиток
// R = 2*W + 2*H - 4, B = (W-2) * (H-2)
fn solve(r: i32, b: i32) -> (i32, i32) {
    let sum = (r + 4) / 2;

    let mut d = 1;
    while d * d <= b {
        if b % d != 0 {
            d += 1;
            continue;
        }

        // Проверяем оба варианта: (W-2, H-2) = (d, B/d) и (B/d, d)
        let candidates = [b / d + 2, d + 2];
        for &w in &candidates {
            let h = sum - w;
            if (w - 2) * (h - 2) == b && w >= h {
                return (w, h);
            }
        }

        d += 1;
    }

    (0, 0)
}







